### ISSUE TITLE
Breeze API's get_command function should not take the current state

### ISSUE THREAD

**Issue by TomerFi** [TomerFi/aioswitcher#551]
### What happened?

I missed [this](https://github.com/TomerFi/aioswitcher/blob/bee61e80fc9fce8e360cf19df1fc4040189c46ea/src/aioswitcher/api/__init__.py#L753) in my reviews before release 3.x.x.
[This](https://github.com/TomerFi/aioswitcher/blob/bee61e80fc9fce8e360cf19df1fc4040189c46ea/src/aioswitcher/api/__init__.py#L753) should be fixed ASAP and I think we should recommend people to avoid using the Breeze API until we fix this to avoid breaking changes.

I think it's a poor practice to require users to do a `get_state` before they do a `get_command` so that we have the current state of the device for toggle devices, we can, and should handle the `get_state` ourselves within `get_command`, much like we do for functions that require a login operation.

```python
def get_command(
        self,
        state: DeviceState,
        mode: ThermostatMode,
        target_temp: int,
        fan_level: ThermostatFanLevel,
        swing: ThermostatSwing,
        current_state: DeviceState,
    ) -> SwitcherBreezeCommand:
        """Build command that controls the Breeze device."""
```

### Module Version

0.0.0

### Device Type

Switcher Breeze

### Firmware Version

0.0.0

**Comment by TomerFi** [comment]
@jafar-atili @thecode 
Any ideas on how to (or time) to handle this?

**Comment by thecode** [comment]
**I disagree this needs to be changed.**

The `get_command` should not do any external API. This is just a function for creating a device command from human parameters, In theory user can store commands and send them to change device states without first requiring interaction with the device.

The `current_state` is needed for toggle device AC types, but it does not mean it must be read from the device. For now I implemented the PR in HA to read it from the device, but I am thinking about using the state HA already knows about the device since HA state is more reliable than the device state (after power-cycle device might be wrong about the state, but HA does remember the state).

I am assuming you are mostly worried about `web_api` users, they should have the current state of the device, so they don't have to pull it from the device, in addition, if you don't have a toggle type AC (unfortunately I do have, but it is only 15 models out of few hundreds) you can ignore this parameter.

If we go with your suggestion, why stop at the `get_command`? why not incooperate all of the logic into `control_breeze_device`?

**Comment by TomerFi** [comment]
> The `get_command` should not do any external API.

You are right, but it uses data retrieved from an external API to do its work.

> This is just a function for creating a device command

Yes, because it's designed that way, it could may as well have been designed as part of the API and not as an helper function.

> In theory user can store commands and send them to change device states without first requiring interaction with the device.

I think that's a bad practice, if you store a command to use later, considering toggle devices, there's a 50/50 chance you have the wrong _current state_.

> The current_state is needed for toggle device AC types, but it does not mean it must be read from the device.

That was my initial thought, that was the reason I was surprised to see that this function requires the current state, if we don't care if the current state we get is the actual state, why do we need it at all? can't we just use the desired state?

> For now I implemented the PR in HA to read it from the device, but I am thinking about using the state HA already knows about the device since HA state is more reliable than the device state (after power-cycle device might be wrong about the state, but HA does remember the state).

Great design, for Home Assistant, but for other clients, clients that do not store the previous state in the memory, i.e. switcher_webapi, this means a requirement for a _get state_ prior to every _get command_.

> I am assuming you are mostly worried about web_api users

Hehe, you know me. 😁 

> if you don't have a toggle type AC (unfortunately I do have, but it is only 15 models out of few hundreds) you can ignore this parameter.

That's a good point,
as the _current state_ arg is optional, then users who doesn't have the current state, can use just the desired state.
If that's the case, then by all means, lets keep the function as is, maybe just set a default value and clarify the docs a bit.

But, although I know the answer and I'm sure it works well, do we have a way to make sure that toggle devices that passes a "wrong state" won't get an error?
i.e. turning on an already turned-on device, "works as expected" (just like a real remote).

> If we go with your suggestion, why stop at the get_command? why not incooperate all of the logic into control_breeze_device?

That is also a good point,
if we end up removing the need for the _current state_ all together, then you are right, this would 
probably be better combined with the control_breeze_device function.

---

To sum it up, I think our options are:
1. Modify the _get command_ function and remove the need for the _current state_ incorporating it as part of the logic in the _control breeze device_ function.

2. Keep the _get command_ as-is, while clarifying the fact that the _current state_ arg is optional for users who write applications with ability to store the current state in memory, or don't mind shooting a _get state_ out first.

You would think I, as the one who opened this issue, will prefer option 1, but no, actually you talked me into option 2, if possible. 😄 

What is your take on this?

**Comment by jafar-atili** [comment]
@TomerFi, so just to confirm, you are suggesting something like the implementation below?: 
https://github.com/TomerFi/aioswitcher/blob/bee61e80fc9fce8e360cf19df1fc4040189c46ea/scripts/control_device.py#L259

The user just call it partially with the parameter/s they'd like to change and the function should figure out the rest of them based on the state?

That's ok, I think we can add another function to achieve it, but I think the `get_command` grants more control and flexibility.

**Comment by TomerFi** [comment]
> so just to confirm, you are suggesting something like the implementation below

@jafar-atili  Exactly, thank you! 😃 

But, after reading @thecode comments pointing out we're verifying the _current state_ exists, then I don't mind leaving everything as-is, and I can just clarify the documentation.

**Comment by thecode** [comment]
> The user just call it partially with the parameter/s they'd like to change and the function should figure out the rest of them based on the state?
> 
> That's ok, I think we can add another function to achieve it, but I think the `get_command` grants more control and flexibility.

In this case I think the best would be:
1. Keep the `get_command` as it is now, but the usage of it will me more like a low level command (still accessible, not private)
2. Change the name of `control_thermostat` to `send_thermostat_command` (if you have a better name feel free to do), do not change the API, again this will be more of a low level command.
3. Create new `control_thermostat` function that has all parameters with defaults to `None` which will do all the logic, so that the user can only supply the parameters he wants to change and the function will internally do the logic for combining the user supplied states with the device states.

What do you think?

**Comment by TomerFi** [comment]
@thecode 
Agreed, no breaking changes, I'll work on this and push a PR tonight.

**Comment by thecode** [comment]
> @thecode Agreed, no breaking changes, I'll work on this and push a PR tonight.

A small breaking change because we change the `control_thermostat` and the old one will be `send_thermostat_command`, but I think this is acceptable for this stage, I don't believe it is used anywhere yet.

You can find a similar code that implements this here:
https://github.com/home-assistant/core/blob/0314687e3a91aa9b84db3cd675cf11d07dd54c28/homeassistant/components/switcher_kis/climate.py#L164-L200
so basically this logic will move from the integration into `aioswitcher`

**Comment by jafar-atili** [comment]
> @thecode
> Agreed, no breaking changes, I'll work on this and push a PR tonight.

I just noticed [this](https://github.com/TomerFi/aioswitcher/blob/bee61e80fc9fce8e360cf19df1fc4040189c46ea/scripts/control_device.py#L280) leftover comment which says ` # First time it'll download the IRSet JSON file from switcher`, I think it needs to be removed

**Comment by TomerFi** [comment]
I can remove it with my PR tonight.

**Comment by thecode** [comment]
While writing my comment https://github.com/TomerFi/aioswitcher/issues/551#issuecomment-1250209242 I realized this is indeed the correct approach as it moves the device logic upstream (`aioswitcher`) and does not require the user to do these steps.

Please note that we need to do a similar to the `get_swing_command`
https://github.com/TomerFi/aioswitcher/blob/bee61e80fc9fce8e360cf19df1fc4040189c46ea/src/aioswitcher/api/__init__.py#L725

We have two options to implement this:
1. Incorporate this logic also into `control_thermostat`, this will require the function to check if a swing parameter is supplied and the device uses `separated_swing_command` and will send an additional swing mode change command (it can't be combined with the other device control commands as it is a separate command.
2. Add another function `control_thermostat_swing` which will be a very simple one, just get the command and send it to the device. In this case we don't need to get the current state from the device since `separated_swing` is stateless (device does not know the state of this type of swing.

**Comment by jafar-atili** [comment]
> > @thecode Agreed, no breaking changes, I'll work on this and push a PR tonight.
> 
> A small breaking change because we change the `control_thermostat` and the old one will be `send_thermostat_command`, but I think this is acceptable for this stage, I don't believe it is used anywhere yet.
> 
> You can find a similar code that implements this here: https://github.com/home-assistant/core/blob/0314687e3a91aa9b84db3cd675cf11d07dd54c28/homeassistant/components/switcher_kis/climate.py#L164-L200 so basically this logic will move from the integration into `aioswitcher`

I agree, its lovely the API gives you the option to control multiple parameters in one command, but in 99% of the cases, this API will be fired always with single change as this is the way most of the GUI's designed (you don't change multiple params and then click Apply), you expect the change to happen following your change in the GUI.
